internal/pkg/policy/monitor.go

```diff
@@ -0,0 +1,351 @@
+	policies map[string]policyT
+
+	policyF       policyFetcher
+	policiesIndex string
+	throttle      time.Duration
+}
+
+// Output returns a new policy that needs to be sent based on the current subscription.
+func (s *subT) Output() <-chan model.Policy {
+	return s.c
+}
+
+// NewMonitor creates the policy monitor for subscribing agents.
+func NewMonitor(bulker bulk.Bulk, monitor monitor.Monitor, throttle time.Duration) Monitor {
+	return &monitorT{
+		log:           log.With().Str("ctx", "policy agent manager").Logger(),
+		bulker:        bulker,
+		monitor:       monitor,
+		kickCh:        make(chan struct{}),
+		policies:      make(map[string]policyT),
+		throttle:      throttle,
+		policyF:       dl.QueryLatestPolicies,
+		policiesIndex: dl.FleetPolicies,
+	}
+}
+
+// Run runs the monitor.
+func (m *monitorT) Run(ctx context.Context) error {
+	s := m.monitor.Subscribe()
+	defer m.monitor.Unsubscribe(s)
+
+LOOP:
+	for {
+		select {
+		case <-ctx.Done():
+			break LOOP
+		case <-m.kickCh:
+			if err := m.process(ctx); err != nil {
+				return err
+			}
+		case hits := <-s.Output():
+			policies := make([]model.Policy, len(hits))
+			for i, hit := range hits {
+				err := hit.Unmarshal(&policies[i])
+				if err != nil {
+					return err
+				}
+			}
+			if err := m.processPolicies(ctx, policies); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (m *monitorT) process(ctx context.Context) error {
+	policies, err := m.policyF(ctx, m.bulker, dl.WithIndexName(m.policiesIndex))
+	if err != nil {
+		return err
+	}
+	if len(policies) == 0 {
+		m.log.Debug().Msg("no policy to monitor")
+		return nil
+	}
+	return m.processPolicies(ctx, policies)
+}
+
+func (m *monitorT) processPolicies(ctx context.Context, policies []model.Policy) error {
+	if len(policies) == 0 {
+		// nothing to do
+		return nil
+	}
+	latest := m.groupByLatest(policies)
+	for _, policy := range latest {
+		if err := m.rollout(ctx, policy); err != nil {
+			if err == context.Canceled {
+				return err
+			}
+			return fmt.Errorf("failed rolling out policy %s: %w", policy.PolicyId, err)
+		}
+	}
+	return nil
+}
+
+func (m *monitorT) groupByLatest(policies []model.Policy) map[string]model.Policy {
+	latest := make(map[string]model.Policy)
+	for _, policy := range policies {
+		curr, ok := latest[policy.PolicyId]
+		if !ok {
+			latest[policy.PolicyId] = policy
+			continue
+		}
+		if policy.RevisionIdx > curr.RevisionIdx {
+			latest[policy.PolicyId] = policy
+			continue
+		} else if policy.RevisionIdx == curr.RevisionIdx && policy.CoordinatorIdx > curr.CoordinatorIdx {
+			latest[policy.PolicyId] = policy
+		}
+	}
+	return latest
+}
+
+func (m *monitorT) rollout(ctx context.Context, policy model.Policy) error {
+	zlog := m.log.With().Str("policyId", policy.PolicyId).Logger()
+
+	subs := m.updatePolicy(policy)
+	if subs == nil {
+		return nil
+	}
+	if len(subs) == 0 {
+		zlog.Info().Msg("no pending subscriptions to revised policy")
+		return nil
+	}
```

Comment:
can replace both ifs above with one 
```
if len(subs) == 0 {
```

Reply:
It could but they have a different meaning. `nil` means the policy should not be rolled out because the coordinatorIdx is 0, and `len(0)` means that no one is subscribed to it